### PR TITLE
Exactly-once message delivery

### PR DESCRIPTION
**Pull request by SzymonPobiega**

### CODE REVIEW

File: content/posts/exactly-once-delivery.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Exactly-once message delivery
+
+It is widely known that exactly-once message delivery is impossible in any distributed system. But what exactly is exactly-once delivery? To answer that question we need to first ask what do we understand as message delivery. This is not an easy task. The receiving system is not a single blob of code. It consists of multiple layers. Is the message considered "delivered" when all its bytes are read from the network cable? If not then maybe when the message is passed to the application? Or maybe when the application completes the procedure that acted on the data? Finally, maybe we should consider a message delivered when the TCP connection is was received on is closed?
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md]: could we link to http://web.mit.edu/Saltzer/www/publications/endtoend/endtoend.pdf

File: content/posts/exactly-once-delivery.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+As odd as the last option might sound, this is the only option in which the sender application can be sure that the receiver got the message. As you can see, the term "message delivery" can be confusing. This is why throughout this series of blog post we used the term "message processing" instead. The has much more clear semantics. A message is processed when all side-effects of the procedure used to handle that message are durably persisted.
+
+Based on that we can establish some derived terms. *At-most-once* message processing means that for each message sent, the side effects of its processing may either be missing or applied once. In contrast *at-least-once* message processing means that the side effects ma be applied multiple times but they may never be skipped. Finally, exactly-once message processing means that the side effects are guaranteed to be applied once.
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md]: >
```suggestion
Based on that we can establish some derived terms. *At-most-once* message processing means that for each message sent, the side effects of its processing may either be missing or applied once. In contrast *at-least-once* message processing means that the side effects may be applied multiple times but they may never be skipped. Finally, exactly-once message processing means that the side effects are guaranteed to be applied once.
```

File: content/posts/exactly-once-delivery.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+### How
+
+Contrary to a popular opinion, *exactly-once* message processing as defined above it possible in real world. It is really unfortunate that the Internet is full of claims according to which *exactly-once delivery* is impossible because of the so-called [two-generals problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem). What that problem *really* illustrates is that *consensus* is impossible in the asynchronous network in which arbitrary message loss is possible. So let's focus on the problem at hand. How do we achieve *exactly-once* message processing in a real-life distributed system?
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md]:
```suggestion
Contrary to popular opinion, *exactly-once* message processing as defined above it possible in real-world systems. Even more importantly in the context of distributed, transactional systems it's significantly more important than the more discussed *exactly-once delivery*.

It is really unfortunate that the Internet is full of distracting claims about *exactly-once delivery* linked to the so-called [two-generals problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem).  A thought experiment illustrating that *consensus* is impossible in the asynchronous network in which arbitrary message loss is possible. So let's focus on the real problem at hand. How do we achieve *exactly-once* message processing in a real-life distributed system?
```

File: content/posts/exactly-once-delivery.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+### Layers
+
+The concept of layer is very useful when describing software systems. Layers allow us to contain the complexity of solving certain class of problems while exposing simple abstractions to the layers above. We will now look at ways of providing *exactly-once* processing guarantees working our way up the stack from the lowest layers to the highest.
+
+### Message queue
+
+The most basic way for two processes can communicate is to send sequences of bytes to each other directly. For the sake of clarity in this article we will use a different model in which processes send messages via a *message queue*. All we need to know about the *queue* is that it can durably store messages and allows retrieving messages in First-In, First-Out (FIFO) order.
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md]: Do we need the FIFO property?

File: content/posts/exactly-once-delivery.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+Based on that we can establish some derived terms. *At-most-once* message processing means that for each message sent, the side effects of its processing may either be missing or applied once. In contrast *at-least-once* message processing means that the side effects may be applied multiple times but they may never be skipped. Finally, exactly-once message processing means that the side effects are guaranteed to be applied once.
+
+### How
+
+Contrary to a popular opinion, **exactly-once message processing is possible** in real world. It is really unfortunate that the majority of the content on the Internet still uses the *delivery*, not *processing* model. It is even worse that most of that content claims *exactly-once delivery* is impossible due to the so-called [two-generals problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem). What that problem *really* illustrates is that *consensus* is impossible in the asynchronous network in which arbitrary message loss is possible.
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md]:
```suggestion
Contrary to a popular opinion, **exactly-once message processing is possible** in the real world. It is really unfortunate that the majority of the content on the Internet focuses on *delivery* and not *processing* which far more significant in transactional systems. It is even worse than a significant part of that content claims *exactly-once delivery* is impossible as proven by the [two-generals problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem). A thought experiment demonstrating *consensus* impossibility in an asynchronous network. 
```

File: content/posts/exactly-once-delivery.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+The most basic way for two processes can communicate is to send sequences of bytes to each other directly. For the sake of clarity in this article we will use a different model in which processes send messages via a *message queue*. All we need to know about the *queue* is that it can durably store messages and allows retrieving messages in First-In, First-Out (FIFO) order.
+
+### Transactions
+
+When we defined message processing the key concept we used was durable persistence of side effects. We did not specify, though, how and where these side effects are persisted. These side effects come in two flavours. First, they include changes in the data store used by the receiving process to store its state. Second, the side effects are also all the messages sent out as a result of processing the incoming messages.
+
+In order to guarantee *exactly-once* processing we can use the transactions that include three operations:
+ - take a message off the queue
+ - modify the application state
+ - enqueue resulting messages
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md:34]: Is there "- remove the message from the input queue" part?

File: content/posts/exactly-once-delivery.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+Transactions guarantee that included operations are atomic and durable. That means that either all three operations are completed or none is. Also, once a transaction is accepted by a system, the system guarantees it won't disappear in future. Transactions are implemented using a concept of a log -- an ordered sequence of operations stored on disk. When transaction is submitted, its description is first written to a log. Only after the log is made durable (e.g. by writing a checksum and ensuring that all write buffers are flushed) are the operations actually applied to the data.
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md:37]: Should we add the isolation aspect here as well and say that "produce the side-effects" should be serialized over state it uses?

File: content/posts/exactly-once-delivery.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+One option is to extend the concept of transaction. If transaction within one resource are useful, surely transactions that span multiple resources would be even more useful. At least that's what people thought in the early 80s when they came up with the concept of distributed transactions. Here we meet our two generals again. Remember? We mentioned that the two-generals problem shows that distributed consensus is not possible. Well, if it is not possible then how do people claim it works? It turns out the impossibility proof is based on the assumption of an asynchronous network in which messages may be delayed arbitrarily long. In such a network no consensus algorithm can guarantee progress. Fortunately real-world networks are not asynchronous. They are more like *semi-asynchronous networks* and that means that messages are delivered eventually. This seemingly weak guarantee is enough to allow a number of consensus algorithms to be proven reliable (e.g. Paxos and Raft).
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md]: Should we add link to https://pdos.csail.mit.edu/6.824/papers/rstar.pdf

Comment (tmasternak) [inline content/posts/exactly-once-delivery.md]: Should we unbox the "consensus is impossible" a bit more and maybe say:

> Impossibility of consensus (assuming deterministic algorithm and asynchronous network) means that there can't be an algorithm that ensures consensus is met under all possible network conditions. In other words, it states that no algorithm can guarantee consensus **always**. That does not imply that in vast majority of real-world conditions consensus will be met. This is exactly why Paxos, Raft or Viewstamped Replication algorithms work (under the common definition of the word). Most of the time they reach consensus but in some network failure conditions, they will fail short.

File: content/posts/exactly-once-delivery.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+So what is that consensus and how does it work? It can be defined in a number of ways but for our purposes it means that for each transaction two (or more) nodes of a system agree that is has been accepted or rejected. There are two widely known types of consensus algorithms. One type is represented by Paxos and Raft we mentioned before. These algorithms are used by distributed databases to ensure data consistency across nodes. The other type is represented by the infamous [Two-Phase Commit](https://exactly-once.github.io/posts/notes-on-2pc/) (sometimes referred to as 2PC). These algorithms are meant to coordinate transactions between different (usually heterogeneous) data stores. For the purpose of *exactly-once* delivery this second type is more relevant. We'll explain it using the implementation provided by Microsoft in form of Distributed Transactions Coordinator (DTC) service.
```
Comment (tmasternak) [inline content/posts/exactly-once-delivery.md:49]: Technically 2PC is different and harder than consensus. Consensus does not assume that any of the nodes can put veto to the value proposed. In other words, these are families of algorithms i.e. consensus and atomic commit protocols.

E.g. there is [Paxos Commit](https://www.microsoft.com/en-us/research/publication/consensus-on-transaction-commit/) that implements atomic commit protocol over Paxos

File: content/posts/exactly-once-delivery.md (hunk 7 above)
Comment (SzymonPobiega) [reply content/posts/exactly-once-delivery.md:37]: I think that's a different aspect. But I can mention in the beginning that all these solutions assume single threaded processing. If you add multiple threads, you need to add concurrency control to ensure correctness.

File: content/posts/exactly-once-delivery.md (hunk 8 above)
Comment (SzymonPobiega) [reply content/posts/exactly-once-delivery.md]: I reworded this section slightly.

File: content/posts/exactly-once-delivery.md (hunk 9 above)
Comment (SzymonPobiega) [reply content/posts/exactly-once-delivery.md:49]: Thanks for this one! I missed the difference ❤️

**Comment by SzymonPobiega** [comment]
@tmasternak I think I applied all your suggestions. Thanks for the review ❤️ I learned a lot!

**Pull request merged**